core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java

```diff
@@ -451,24 +452,48 @@ private RelationPlan createTableCreationPlan(Analysis analysis, Query query)
         }
 
         ConnectorTableMetadata tableMetadata = create.getMetadata().orElseThrow();
-
         Optional<TableLayout> newTableLayout = create.getLayout();
 
-        List<String> columnNames = tableMetadata.getColumns().stream()
-                .filter(column -> !column.isHidden()) // todo this filter is redundant
-                .map(ColumnMetadata::getName)
-                .collect(toImmutableList());
+        List<Symbol> visibleFieldMappings = getVisibleFieldMappings(plan);
 
         String catalogName = destination.getCatalogName();
         CatalogHandle catalogHandle = metadata.getCatalogHandle(session, catalogName)
                 .orElseThrow(() -> semanticException(CATALOG_NOT_FOUND, query, "Destination catalog '%s' does not exist", catalogName));
+
+        Assignments.Builder assignments = Assignments.builder();
+        ImmutableList.Builder<ColumnMetadata> finalColumnsBuilder = ImmutableList.builder();
+
+        forEachPair(tableMetadata.getColumns().stream(), visibleFieldMappings.stream(), (column, fieldMapping) -> {
+            Expression expression;
+            Type tableType = column.getType();
+            Type queryType = symbolAllocator.getTypes().get(fieldMapping);
+            if (queryType.equals(tableType) || typeCoercion.isTypeOnlyCoercion(queryType, tableType)) {
+                expression = fieldMapping.toSymbolReference();
+            }
+            else {
+                expression = noTruncationCast(fieldMapping.toSymbolReference(), queryType, tableType);
+            }
+            Symbol output = symbolAllocator.newSymbol(column.getName(), column.getType());
+            assignments.put(output, expression);
+            finalColumnsBuilder.add(column);
+        });
```

Comment:
How is this different from what's being done in INSERT planning?

Reply:
Then, why is it structurally so different? If both share the same (complex) logic, we should factor out and reuse. Or, at a minimum the code should look the same across both implementations.

Reply:
I don't believe this comment has been addressed.

Reply:
The only part that is actually reusable in both pieces of code is:

```
if (queryType.equals(tableType) || typeCoercion.isTypeOnlyCoercion(queryType, tableType)) {
    expression = fieldMapping.toSymbolReference();
}
else {
    expression = noTruncationCast(fieldMapping.toSymbolReference(), queryType, tableType);
} 
```

The rest is quite different as the INSERT needs to deal with missing columns and hidden columns. Both of them are not relevant in the CTAS use case.

I extracted that logic and use it in both instances.